monitoring/ping.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+	"log"
+
+	"github.com/gravitational/satellite/agent"
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/trace"
+	serf "github.com/hashicorp/serf/client"
+	ping "github.com/sparrc/go-ping"
+)
+
+const (
+	PingCheckerID = "ping-checker"
+)
+
+func NewPingChecker(serfRPCaddr string, role string) health.Checker {
+	return &PingChecker{
+		serfRPCAddr: serfRPCaddr,
+		role:        role,
+	}
+}
+
+// PingChecker is a checker that verifies if ping between Master Nodes and other
+// ones is below a specified threshold
+type PingChecker struct {
+	serfRPCAddr string
+	role        string
+}
+
+// Name returns the check name
+// Implements health.Checker
+func (c *PingChecker) Name() string {
+	return PingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *PingChecker) Check(ctx context.Context, r health.Reporter) {
+	RttThreshold := 100     // ms
+	slidingWindowSize := 10 // number of ping results to consider per iteration
+
+	// set Probe to be running
+	r.Add(&pb.Probe{
+		Checker: c.Name(),
+		Status:  pb.Probe_Running,
+	})
+
+	// fetch serf config and intantiate client
+	clientConfig := serf.Config{
+		Addr: c.serfRPCAddr,
+	}
+	client, err := serf.ClientFromConfig(clientConfig)
+	if err != nil {
+		return // nil, trace.Wrap(err, "failed to connect to serf")
+	}
+	defer client.Close()
+
+	// only run PingCheck between masters
+	role := c.role
+	if role != agent.RoleMaster {
```

Comment:
Instead of checking this internally, why just not make sure that this probe is added only on masters? I.e. do this on the caller side - where checkers are initialized. Otherwise this probe will be added to all regular nodes and will be firing up without actually doing anything.

Reply:
Good point.
I am actually already doing this by only calling it in the `cmd/agent/checkers.go:addToMasters()` function ... so I guess that's just redundant.